### ISSUE TITLE
Scope

### ISSUE THREAD

**Issue by rjvim** [nobejs/discuss#1]
* [x] User can post a query (Anonymous)
* [x] User can see the queries they posted
* [x] User can delete the query
* [x] User can edit the query

### Query

- A query can multiple tags
- Query has a title
- Query has a body
- Query has a owner
- Query has a meta data object (json object)

* [x] Architect DB Schema
* [x] Brainstorm on how to make browsing faster

----

* [x] A user can post an answer (Option: Anonymous)
* [x] A user can edit answer
* [x] A user can delete answer
* [x] A user can see all answers for a query

* [x] Architect DB Schema
* [x] A user can see all his list of answers

----

* [x] Admin can create tags
* [x] Admin can edit tags
* [x] Admin can view tags
* [x] Admin can delete tags

----

Browse all the queries

* [ ] Filter by title
* [x] Filter by user
* [x] Filter by tags
* [ ] Filter by number of answers
* [x] Sort by created_at, updated_at, last_answered_at

**Comment by rohan099** [comment]
**Architect DB Schema**

tables

**query**

- uuid 
- user_uuid
- data(JSONB):{
   - question
   - country
   - response_period(number of days)
   - anonymous(true/false)
}
- createdAt
- updatedAt
- deletedAt

**tags**

- uuid
- name


**query_tags**  (pivot table)

- uuid
- query_uuid
- tags_uuid


 @rjvim @rajeshn95  please check this

**Comment by rjvim** [comment]
@rohan099 Query DB is not covering some aspects:

If we think of everything in terms of jsonb, it would be difficult to establishing ground rules like, title is mandatory etc., 

So, title and body would become columns, meta is an object which can store more data which is provided as feature by "discuss" 

I think it would more on following lines (I am excluding timestamps):

| uuid | title        | body     | owner_id                                     | anonymous  | meta              |
|------|--------------|----------|----------------------------------------------|------------|-------------------|
|      | not nullable | nullable | user's uuid, but this column would be string | true/false | {} - jsonb object |


Tags and Queries seem fine, I think uuid can be dropped and table/column names can be more consistent:

1. queries
2. tags
3. queries_tags

queries_tags can have following columns:

- query_uuid
- tag_uuid

Basically, (query_uuid, tag_uuid) become the primary key together, usually this can be achieved in knexjs like following:

```
exports.up = function (knex) {
  return knex.schema.alterTable("template_versions", function (table) {
    table.unique(["template_uuid", "version"]);
  });
};
```

**Comment by rohan099** [comment]
@rjvim got it

**Comment by rohan099** [comment]
**query**

| uuid  | title |  body | owner_uuid | anonymous | meta | timestamps |
| ------------- | ------------- |------------- |------------- |------------- |------------- |------------- |
|  UUID  | text(required)  | text(nullable)  | string (required) | Boolean (required) | JSONB ({country, response_period}) | createdAt, updatedAt

**tags**

| uuid  | name |
| ------------- | ------------- |
| uuid  | string(required)  |

**queries_tags**

| query_uuid  | tag_uuid |
| ------------- | ------------- |
| (required)  | (required)  |

**Comment by rohan099** [comment]
@rjvim  please check this DB schema

**Responses**

| uuid  | query_uuid |  owner_uuid |answer |  anonymous | meta | timestamps |
| ------------- | ------------- |------------- |------------- |------------- |------------- |------------- |
|  UUID  | string(required)  | string(required)  | text (required) | Boolean (required) | JSONB nullable (any attachment or likes(future scope) | createdAt, updatedAt

**Comment by rjvim** [comment]
@rohan099 Seems fine.

**Comment by rohan099** [comment]
@rjvim  here we are creating queries and and responses but we also need the relation with creator mainly we need to show the name and contact info in wiredUP app, we have owner_uuid in table, how can we add that data in api call or we should do it from frontend using owner_uuid.

we can add that profile data in meta here?

**Comment by rjvim** [comment]
@rohan099 Discuss service itself doesn't have users, so it can't be augmented with such information. And it's not responsibility of Discuss service. It will only have user_uuid, using which the main application has to decide how to get profile information like name, email, profile pic etc.,

**Comment by rohan099** [comment]
ok @rjvim
